src/Cooked/MockChain/BlockChain.hs

```diff
@@ -203,7 +204,30 @@ resolveDatum out =
           (out ^. outputReferenceScriptL)
     PV2.NoOutputDatum -> return Nothing
 
--- | Try to resolve the validator that owns an output: If the output is owned by
+-- | Like 'resolveDatum', but also tries to use 'fromBuiltinData' to extract a
+-- datum of the suitable type.
+resolveTypedDatum ::
+  ( IsAbstractOutput out,
+    ToOutputDatum (DatumType out),
+    MonadBlockChainBalancing m,
+    PV2.FromData a
+  ) =>
+  out ->
+  m (Maybe (ConcreteOutput (OwnerType out) a (ValueType out) (ReferenceScriptType out)))
+resolveTypedDatum out = do
+  mOut <- resolveDatum out
+  case mOut of
+    Nothing -> return Nothing
+    Just out' ->
+      return $
+        ConcreteOutput
+          <$> Just (out' ^. outputOwnerL)
+          <*> Just (out' ^. outputStakingCredentialL)
+          <*> Just (out' ^. outputValueL)
+          <*> (case out' ^. outputDatumL of PV2.Datum datum -> PV2.fromBuiltinData datum)
```

Comment:
Are there other patterns? If so, I think we should have a runtime error for those, if not, we should use a let.